# Prosper Loan Data Exploration
## by Jeffrey Van Anderson

## Preliminary Wrangling

**Introduction:**
> This document explores data from 113,936 loans from "Prosper" and an extensive 81 variables.  Univariate, bivariate and finally multivatiate visualizations are used as the primary exploritory tools here.  Additionally, only fifteen variables are used to address questionns in this exploration.

**Step 1:** First, all the necessary packages are imported and `*.csv` data file is imported as pandas dataframe

In [1]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [2]:
#import *.csv file as pandas dataframe and display first five lines
df = pd.read_csv('prosperLoanData.csv')
df.head()

ListingKey  ListingNumber            ListingCreationDate  \
0  1021339766868145413AB3B         193129  2007-08-26 19:09:29.263000000   
1  10273602499503308B223C1        1209647  2014-02-27 08:28:07.900000000   
2  0EE9337825851032864889A          81716  2007-01-05 15:00:47.090000000   
3  0EF5356002482715299901A         658116  2012-10-22 11:02:35.010000000   
4  0F023589499656230C5E3E2         909464  2013-09-14 18:38:39.097000000   

  CreditGrade  Term LoanStatus           ClosedDate  BorrowerAPR  \
0           C    36  Completed  2009-08-14 00:00:00      0.16516   
1         NaN    36    Current                  NaN      0.12016   
2          HR    36  Completed  2009-12-17 00:00:00      0.28269   
3         NaN    36    Current                  NaN      0.12528   
4         NaN    36    Current                  NaN      0.24614   

   BorrowerRate  LenderYield  ...  LP_ServiceFees  LP_CollectionFees  \
0        0.1580       0.1380  ...         -133.18                0.0   
1        0.0920       0.0820  ...            0.00                0.0   
2        0.2750       0.2400  ...          -24.20                0.0   
3        0.0974       0.0874  ...         -108.01                0.0   
4        0.2085       0.1985  ...          -60.27                0.0   

   LP_GrossPrincipalLoss  LP_NetPrincipalLoss LP_NonPrincipalRecoverypayments  \
0                    0.0                  0.0                             0.0   
1                    0.0                  0.0                             0.0   
2                    0.0                  0.0                             0.0   
3                    0.0                  0.0                             0.0   
4                    0.0                  0.0                             0.0   

   PercentFunded  Recommendations InvestmentFromFriendsCount  \
0            1.0                0                          0   
1            1.0                0                          0   
2            1.0                0                          0   
3            1.0                0                          0   
4            1.0                0                          0   

  InvestmentFromFriendsAmount Investors  
0                         0.0       258  
1                         0.0         1  
2                         0.0        41  
3                         0.0       158  
4                         0.0        20  

[5 rows x 81 columns]

**Step 2:** Next, information is displayed about the dataframe in order to become familiar with the data

In [3]:
#assess data types and dataframe shape
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 81 columns):
ListingKey                             113937 non-null object
ListingNumber                          113937 non-null int64
ListingCreationDate                    113937 non-null object
CreditGrade                            28953 non-null object
Term                                   113937 non-null int64
LoanStatus                             113937 non-null object
ClosedDate                             55089 non-null object
BorrowerAPR                            113912 non-null float64
BorrowerRate                           113937 non-null float64
LenderYield                            113937 non-null float64
EstimatedEffectiveYield                84853 non-null float64
EstimatedLoss                          84853 non-null float64
EstimatedReturn                        84853 non-null float64
ProsperRating (numeric)                84853 non-null float64
ProsperRating (Alpha) 

**Step 3:**  After briefly previewing the data, it is clear that there are excessive column variables to address in this one report.  A select few are queried and applied to a new dataframe for analysis.

In [4]:
#create list of collumns to keep for analysis
cols_to_keep = ['ListingKey', 'LoanStatus', 'BorrowerAPR', 'BorrowerRate', 'ListingCategory (numeric)',
                'BorrowerState', 'EmploymentStatus', 'EmploymentStatusDuration',
                'IsBorrowerHomeowner', 'StatedMonthlyIncome', 'LoanOriginalAmount', 'MonthlyLoanPayment',
                'CurrentDelinquencies', 'AmountDelinquent', 'DelinquenciesLast7Years']

#assign queried collumns as df_trimmed variable
df_trimmed = df[cols_to_keep]

#confirm successful trim
df_trimmed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 15 columns):
ListingKey                   113937 non-null object
LoanStatus                   113937 non-null object
BorrowerAPR                  113912 non-null float64
BorrowerRate                 113937 non-null float64
ListingCategory (numeric)    113937 non-null int64
BorrowerState                108422 non-null object
EmploymentStatus             111682 non-null object
EmploymentStatusDuration     106312 non-null float64
IsBorrowerHomeowner          113937 non-null bool
StatedMonthlyIncome          113937 non-null float64
LoanOriginalAmount           113937 non-null int64
MonthlyLoanPayment           113937 non-null float64
CurrentDelinquencies         113240 non-null float64
AmountDelinquent             106315 non-null float64
DelinquenciesLast7Years      112947 non-null float64
dtypes: bool(1), float64(8), int64(2), object(4)
memory usage: 12.3+ MB


**Step 4:** With the separate dataframe, `df_trimmed` having fifteen columns, a few more functions can be applied to learn more about the data

In [6]:
#review distribution of data in variables
df_trimmed.describe()

BorrowerAPR   BorrowerRate  ListingCategory  \
count  113912.000000  113937.000000    113937.000000   
mean        0.218828       0.192764         2.774209   
std         0.080364       0.074818         3.996797   
min         0.006530       0.000000         0.000000   
25%         0.156290       0.134000         1.000000   
50%         0.209760       0.184000         1.000000   
75%         0.283810       0.250000         3.000000   
max         0.512290       0.497500        20.000000   

       EmploymentStatusDuration  StatedMonthlyIncome  LoanOriginalAmount  \
count             106312.000000         1.139370e+05        113937.00000   
mean                  96.071582         5.608026e+03          8337.01385   
std                   94.480605         7.478497e+03          6245.80058   
min                    0.000000         0.000000e+00          1000.00000   
25%                   26.000000         3.200333e+03          4000.00000   
50%                   67.000000         4.666667e+03          6500.00000   
75%                  137.000000         6.825000e+03         12000.00000   
max                  755.000000         1.750003e+06         35000.00000   

       MonthlyLoanPayment  CurrentDelinquencies  AmountDelinquent  \
count       113937.000000         113240.000000     106315.000000   
mean           272.475783              0.592052        984.507059   
std            192.697812              1.978707       7158.270157   
min              0.000000              0.000000          0.000000   
25%            131.620000              0.000000          0.000000   
50%            217.740000              0.000000          0.000000   
75%            371.580000              0.000000          0.000000   
max           2251.510000             83.000000     463881.000000   

       DelinquenciesLast7Years  
count            112947.000000  
mean                  4.154984  
std                  10.160216  
min                   0.000000  
25%                   0.000000  
50%                   0.000000  
75%                   3.000000  
max                  99.000000

In [7]:
#view entries with null `AmountDelinquent` observations
df_trimmed[df_trimmed['AmountDelinquent'].isnull()].sample(5)

ListingKey LoanStatus  BorrowerAPR  BorrowerRate  \
22088  2B3C3370880726790EC70CD  Completed      0.11696        0.1100   
51995  795B3364998059162D599E4  Defaulted      0.18475        0.1775   
37407  307E3372565949873D70113  Completed      0.08684        0.0800   
35053  3ECE3375323602400DD1F8C  Defaulted      0.29776        0.2900   
57982  9C81337192177363080C081  Completed      0.09476        0.0775   

       ListingCategory BorrowerState EmploymentStatus  \
22088                0           NaN    Not available   
51995                0            FL              NaN   
37407                0           NaN    Not available   
35053                0           NaN    Not available   
57982                0           NaN    Not available   

       EmploymentStatusDuration  IsBorrowerHomeowner  StatedMonthlyIncome  \
22088                       NaN                 True          4833.333333   
51995                       NaN                 True          3416.666667   
37407                       NaN                False          4583.333333   
35053                       NaN                 True          2666.666667   
57982                       NaN                False          4000.000000   

       LoanOriginalAmount  MonthlyLoanPayment  CurrentDelinquencies  \
22088                5000              163.69                   0.0   
51995                7500              270.20                   0.0   
37407                4950              155.12                   0.0   
35053               15000              628.59                   0.0   
57982                1000               31.22                   0.0   

       AmountDelinquent  DelinquenciesLast7Years  
22088               NaN                      0.0  
51995               NaN                      0.0  
37407               NaN                      0.0  
35053               NaN                      0.0  
57982               NaN                      0.0

**Step 5:**  After inspection, there are a few more tidyness and cleanliness issues that need to be addressed as follows:
* `ListingCategory (numeric)` title needs to be shortened
* there is null data in a few columns
* some float datatypes should be changed to integer.
* the `LoanStatus` column should be ordered for clarity
* the `ListingCategory` column values should be replaced with strings to avoid unnecesssary referencing

In [5]:
#remove ' (numeric)' string from `ListingCategory`
df_trimmed = df_trimmed.rename(index=str, columns={"ListingCategory (numeric)": "ListingCategory"})
df_trimmed.head(1)

ListingKey LoanStatus  BorrowerAPR  BorrowerRate  \
0  1021339766868145413AB3B  Completed      0.16516         0.158   

   ListingCategory BorrowerState EmploymentStatus  EmploymentStatusDuration  \
0                0            CO    Self-employed                       2.0   

   IsBorrowerHomeowner  StatedMonthlyIncome  LoanOriginalAmount  \
0                 True          3083.333333                9425   

   MonthlyLoanPayment  CurrentDelinquencies  AmountDelinquent  \
0              330.43                   2.0             472.0   

   DelinquenciesLast7Years  
0                      4.0

In [8]:
#since the following column names have a lot of zeros and a sample of null `AmountDelinquent` observations have many other 
#null values, it is safe to drop these rows.

#create list of columns with nulls
nulls_todrop = ['AmountDelinquent', 'BorrowerState', 'EmploymentStatusDuration']

#drop row observalions where there are nulls in each specified column
for col in nulls_todrop:
    df_trimmed = df_trimmed[df_trimmed[col].notnull()]
    
#ensure success
df_trimmed.info()

<class 'pandas.core.frame.DataFrame'>
Index: 104555 entries, 0 to 113936
Data columns (total 15 columns):
ListingKey                  104555 non-null object
LoanStatus                  104555 non-null object
BorrowerAPR                 104555 non-null float64
BorrowerRate                104555 non-null float64
ListingCategory             104555 non-null int64
BorrowerState               104555 non-null object
EmploymentStatus            104555 non-null object
EmploymentStatusDuration    104555 non-null float64
IsBorrowerHomeowner         104555 non-null bool
StatedMonthlyIncome         104555 non-null float64
LoanOriginalAmount          104555 non-null int64
MonthlyLoanPayment          104555 non-null float64
CurrentDelinquencies        104555 non-null float64
AmountDelinquent            104555 non-null float64
DelinquenciesLast7Years     104555 non-null float64
dtypes: bool(1), float64(8), int64(2), object(4)
memory usage: 12.1+ MB


In [9]:
df_trimmed.head()

ListingKey LoanStatus  BorrowerAPR  BorrowerRate  \
0  1021339766868145413AB3B  Completed      0.16516        0.1580   
1  10273602499503308B223C1    Current      0.12016        0.0920   
3  0EF5356002482715299901A    Current      0.12528        0.0974   
4  0F023589499656230C5E3E2    Current      0.24614        0.2085   
5  0F05359734824199381F61D    Current      0.15425        0.1314   

   ListingCategory BorrowerState EmploymentStatus  EmploymentStatusDuration  \
0                0            CO    Self-employed                       2.0   
1                2            CO         Employed                      44.0   
3               16            GA         Employed                     113.0   
4                2            MN         Employed                      44.0   
5                1            NM         Employed                      82.0   

   IsBorrowerHomeowner  StatedMonthlyIncome  LoanOriginalAmount  \
0                 True          3083.333333                9425   
1                False          6125.000000               10000   
3                 True          2875.000000               10000   
4                 True          9583.333333               15000   
5                 True          8333.333333               15000   

   MonthlyLoanPayment  CurrentDelinquencies  AmountDelinquent  \
0              330.43                   2.0             472.0   
1              318.93                   0.0               0.0   
3              321.45                   4.0           10056.0   
4              563.97                   0.0               0.0   
5              342.37                   0.0               0.0   

   DelinquenciesLast7Years  
0                      4.0  
1                      0.0  
3                     14.0  
4                      0.0  
5                      0.0

In [10]:
#change float dtypes to integer for four columns
vars_to_int = ['EmploymentStatusDuration', 'CurrentDelinquencies', 'AmountDelinquent', 'DelinquenciesLast7Years']
for var in vars_to_int:
    df_trimmed[var] = df_trimmed[var].astype('int')

In [11]:
df_trimmed.info()

<class 'pandas.core.frame.DataFrame'>
Index: 104555 entries, 0 to 113936
Data columns (total 15 columns):
ListingKey                  104555 non-null object
LoanStatus                  104555 non-null object
BorrowerAPR                 104555 non-null float64
BorrowerRate                104555 non-null float64
ListingCategory             104555 non-null int64
BorrowerState               104555 non-null object
EmploymentStatus            104555 non-null object
EmploymentStatusDuration    104555 non-null int32
IsBorrowerHomeowner         104555 non-null bool
StatedMonthlyIncome         104555 non-null float64
LoanOriginalAmount          104555 non-null int64
MonthlyLoanPayment          104555 non-null float64
CurrentDelinquencies        104555 non-null int32
AmountDelinquent            104555 non-null int32
DelinquenciesLast7Years     104555 non-null int32
dtypes: bool(1), float64(4), int32(4), int64(2), object(4)
memory usage: 10.5+ MB


In [12]:
#change `LoanStatus` to categorical ordered data type for interpretation
#first create ordered list
loan_stat_ord = ['Current', 'Cancelled', 'FinalPaymentInProgress', 'Completed', 'Past Due (1-15 days)',
                 'Past Due (16-30 days)', 'Past Due (31-60 days)', 'Past Due (61-90 days)',
                 'Past Due (91-120 days)', 'Past Due (>120 days', 'Chargedoff', 'Defaulted']

#then create an ordered catagorical datatype variable based on the list
loan_stat_var = pd.api.types.CategoricalDtype(ordered = True,
                                              categories = loan_stat_ord)

#then assign the `LoanStatus` column the datatype variable
df_trimmed['LoanStatus'] = df_trimmed['LoanStatus'].astype(loan_stat_var)

#confirm success
df_trimmed.info()

<class 'pandas.core.frame.DataFrame'>
Index: 104555 entries, 0 to 113936
Data columns (total 15 columns):
ListingKey                  104555 non-null object
LoanStatus                  104539 non-null category
BorrowerAPR                 104555 non-null float64
BorrowerRate                104555 non-null float64
ListingCategory             104555 non-null int64
BorrowerState               104555 non-null object
EmploymentStatus            104555 non-null object
EmploymentStatusDuration    104555 non-null int32
IsBorrowerHomeowner         104555 non-null bool
StatedMonthlyIncome         104555 non-null float64
LoanOriginalAmount          104555 non-null int64
MonthlyLoanPayment          104555 non-null float64
CurrentDelinquencies        104555 non-null int32
AmountDelinquent            104555 non-null int32
DelinquenciesLast7Years     104555 non-null int32
dtypes: bool(1), category(1), float64(4), int32(4), int64(2), object(3)
memory usage: 9.8+ MB


In [13]:
#change `ListingCategory` from integer to category string
#first change dtype to string for iteration
df_trimmed['ListingCategory'] = df_trimmed['ListingCategory'].astype('str')

#assign dictionary of category name to number
cat_dict = {'0' : 'Not Available',
            '1' : 'Debt Consolidation',
            '2' : 'Home Improvement', 
            '3' : 'Business',
            '4' : 'Personal Loan',
            '5' : 'Student Use',
            '6' : 'Auto', 
            '7' : 'Other',
            '8' : 'Baby&Adoption',
            '9' : 'Boat',
            '10' : 'Cosmetic Procedure',
            '11' : 'Engagement Ring',
            '12' : 'Green Loans',
            '13' : 'Household Expenses', 
            '14' : 'Large Purchases',
            '15' : 'Medical/Dental',
            '16' : 'Motorcycle',
            '17' : 'RV',
            '18' : 'Taxes',
            '19' : 'Vacation',
            '20' : 'Wedding Loans'}

#loop over column and replace numbers with category strings
for num, string in cat_dict.items():
    df_trimmed['ListingCategory'] = df_trimmed['ListingCategory'].replace(num, string)

#confirm success
df_trimmed['ListingCategory'].value_counts()

Debt Consolidation    58279
Other                 10487
Not Available          7652
Home Improvement       7431
Business               7180
Auto                   2568
Personal Loan          2380
Household Expenses     1995
Medical/Dental         1522
Taxes                   885
Large Purchases         876
Wedding Loans           770
Vacation                768
Student Use             755
Motorcycle              304
Engagement Ring         217
Baby&Adoption           199
Cosmetic Procedure       91
Boat                     85
Green Loans              59
RV                       52
Name: ListingCategory, dtype: int64

**Step 6:** Now that the data is wrangled, it is time to reflect on it prior to exploratory analysis.

### What is the structure of your dataset?

> After some trimming, tidying and wrangling the dataframe to be explored has considerably fewer columns than the original dataset; however, the number of rows are only slightly less since incomplete data was dropped.  There are fifteen columns in the resulting dataframe (14 variables and one key) and 104,555 rows or loan records with no null values.

> For the fourteen variable collumns, there are the following data types:
> * there are eight numeric columns (four ineger and four float)
> * the `LoanStatus` columns is ordered from no delinquency or `Current` to `Defaulted` later interpretation
> * there are three string or categorical columns for state, employment and listing category
> * finally, the `IsBorrowerHomeowner` column is a simple bool answer.

### What is/are the main feature(s) of interest in your dataset?

> Of primary interest is the relationship between loan status and the other thirteen variables attributed to the borrower.  

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

> The `LoanStatus` column will directly be effected by `CurrentDelinquencies`; however, the `AmountDelinquent` and `DelinquenciesLast7Years` might have a different relationship.  Surely interest rates, emploment information will have a significant effect on LoanStatus, but they must be visually explored first.  Finally, it does not seam likely that `BorrowerState` or `ListingCategory` will be insightful at first, but there may be major differences depending on each state or category. 

## Univariate Exploration

> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.

> Make sure that, after every plot or related series of plots, that you
include a Markdown cell with comments about what you observed, and what
you plan on investigating next.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> Your answer here!

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

> Your answer here!

## Bivariate Exploration

> In this section, investigate relationships between pairs of variables in your
data. Make sure the variables that you cover here have been introduced in some
fashion in the previous section (univariate exploration).

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> Your answer here!

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> Your answer here!

## Multivariate Exploration

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

> Your answer here!

### Were there any interesting or surprising interactions between features?

> Your answer here!

> At the end of your report, make sure that you export the notebook as an
html file from the `File > Download as... > HTML` menu. Make sure you keep
track of where the exported file goes, so you can put it in the same folder
as this notebook for project submission. Also, make sure you remove all of
the quote-formatted guide notes like this one before you finish your report!